rohmu/transfer_pool.py

```diff
@@ -0,0 +1,220 @@
+        self.transfers_heap = []
+        self.last_used = time.monotonic()
+        self.max_age = max_age
+
+    def prune_expired(self) -> None:
+        while self.transfers_heap and self.transfers_heap[0].is_expired():
+            heapq.heappop(self.transfers_heap)
+
+    def get(self) -> Optional[TransferCacheItem]:
+        self.last_used = time.monotonic()
+        self.prune_expired()
+        if self.transfers_heap:
+            try:
+                return heapq.heappop(self.transfers_heap)
+            except IndexError:
+                pass
+        return None
+
+    def put(self, transfer: TransferCacheItem) -> None:
+        self.last_used = time.monotonic()
+        if not transfer.is_expired():
+            heapq.heappush(self.transfers_heap, transfer)
+
+    def age(self) -> float:
+        return time.monotonic() - self.last_used
+
+    def is_expired(self) -> bool:
+        return self.age() > self.max_age
+
+
+class _TransferCacheForThreadSafeTransfer(_TransferCache):
+    """This _TransferCache subclass is a simple implementation that works when the transfer is thread-safe.
+
+    In that case we can have a single instance and always return that one.
+
+    We still apply the expiration logic though.
+    """
+
+    def get(self) -> Optional[TransferCacheItem]:
+        self.last_used = time.monotonic()
+        self.prune_expired()
+        if self.transfers_heap:
+            try:
+                # there will always be only one transfer instance at most in this cache type
+                return self.transfers_heap[0]
+            except IndexError:
+                pass
+        return None
+
+    def put(self, transfer: TransferCacheItem) -> None:
+        self.last_used = time.monotonic()
+        if not self.transfers_heap:
+            # we only put 1 item in the "heap" for this transfer cache type
+            self.transfers_heap.append(transfer)
+
+
+_BASE_TRANSFER_INSTANCE_ATTRS = {"config_model", "log", "notifier", "prefix", "stats"}
+_BASE_TRANSFER_ATTRS = {attr for attr in vars(BaseTransfer) if not attr.startswith("__")} | _BASE_TRANSFER_INSTANCE_ATTRS
+
+# pylint: disable=abstract-method,super-init-not-called
+class SafeTransfer(BaseTransfer[StorageModel]):
+    """Helper class that helps the users in finding bugs in their code handling transfers.
+
+    This class prevents any call to a transfer instance that was returned to the pool.
+    It also logs a warning if the instance is garbage collected (returning it to the pool).
+
+    """
+
+    def __init__(self, item: TransferCacheItem, pool_reclaim: Callable[[TransferCacheItem], None]) -> None:
+        self._item = item
+        self._pool_reclaim = pool_reclaim
+        self._done = False
+
+    def __del__(self) -> None:
+        if not self._done:
+            LOG.warning("Transfer not marked as done was garbage collected. Returning it to the pool now.")
+            self.return_to_pool()
```

Comment:
Probably logging only here would be good enough too - not like we really care about keeping the transfer.